libtenzir/builtins/operators/sort.cpp

```diff
@@ -309,8 +312,175 @@ class plugin final : public virtual operator_plugin<sort_operator> {
   }
 };
 
+// -- TQL2 implementation below ------------------------------------------------
+
+struct sort_expression {
+  ast::expression expr = {};
+  bool reverse = {};
+
+  friend auto inspect(auto& f, sort_expression& x) -> bool {
+    return f.object(x).fields(f.field("expr", x.expr),
+                              f.field("reverse", x.reverse));
+  }
+};
+
+struct sort_key {
+  std::vector<series> chunks = {};
+  bool reverse = {};
+};
+
+struct sort_index {
+  size_t slice = {};
+  int64_t event = {};
+};
+
+class sort_operator2 final : public crtp_operator<sort_operator2> {
+public:
+  sort_operator2() = default;
+
+  explicit sort_operator2(std::vector<sort_expression> sort_exprs)
+    : sort_exprs_{std::move(sort_exprs)} {
+  }
+
+  auto name() const -> std::string override {
+    return "tql2.sort";
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    auto events = std::vector<table_slice>{};
+    auto indices = std::vector<sort_index>{};
+    auto sort_keys = std::vector<sort_key>{};
+    sort_keys.reserve(sort_exprs_.size());
+    for (const auto& sort_expr : sort_exprs_) {
+      sort_keys.emplace_back().reverse = sort_expr.reverse;
+    }
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      const auto length = detail::narrow<int64_t>(slice.rows());
+      indices.reserve(indices.size() + slice.rows());
+      for (int64_t i = 0; i < length; ++i) {
+        indices.push_back({
+          .slice = events.size(),
+          .event = i,
+        });
+      }
+      for (size_t i = 0; i < sort_exprs_.size(); ++i) {
+        sort_keys[i].chunks.push_back(
+          eval(sort_exprs_[i].expr, slice, ctrl.diagnostics()));
+      }
+      events.push_back(std::move(slice));
+    }
+    if (indices.empty()) {
+      co_return;
+    }
+    // TODO: If all chunks for a sort key evaluate to the same type, then we can
+    // choose a faster path where we do not need to evaluate the sort key's type
+    // for each row individually. That may be significantly faster.
+    std::ranges::sort(indices, [&](const sort_index& lhs,
+                                   const sort_index& rhs) {
+      for (const auto& sort_key : sort_keys) {
+        const auto& lhs_key = sort_key.chunks[lhs.slice];
+        const auto& rhs_key = sort_key.chunks[rhs.slice];
+        const auto lhs_null = lhs_key.array->IsNull(lhs.event);
+        const auto rhs_null = rhs_key.array->IsNull(rhs.event);
+        if (lhs_null and rhs_null) {
+          continue;
+        }
+        if (lhs_null or rhs_null) {
+          // Nulls last, independent of sort order.
+          return rhs_null;
+        }
+        // FIXME: Properly implement operator<=> for data_view and data in a way
+        // that orders across types. All numeric types should be comparable, and
+        // nulls should always be last.
```

Comment:
This and tests are the two big remaining things to implement for this PR. @raxyte @jachris you already spent a fair amount of time looking at comparisons across types, so I was wondering if this would be a good point to hand the PR over to you.

Reply:
We found that it is not as easy as expected to adapt `data` and `data_view` (and also `record` and `list`). We could get `sort` in with the "slightly broken" order, and return to this at some other point in time.

Reply:
I think that's a valid option. We should still return to it at some point, i.e., definitely write a follow-up story.

Reply:
Another option would actually be to implement the custom sorting locally within this function, with a note to return to it in the future.